docs/yumdownloader.1
.\" yumdownloader
.TH "yumdownloader" "1" "2005 Nov 28" "Gijs Hollestelle" ""
.SH "NAME"
yumdownloader
.SH "SYNOPSIS"
\fByumdownloader\fP [options] package1 [package2...]
.SH "DESCRIPTION"
.PP 
\fByumdownloader\fP is a program for downloading RPMs from Yum repositories.
.PP 
.SH "GENERAL OPTIONS"
.IP "\fB\-h, \-\-help\fP"
Help; display a help message and then quit\&.
.IP "\fB\-c CONFIG, \-\-config=CONFIG\fP"
Specify the config file to use.  Defaults to /etc/yum.conf.
.IP "\fB\-\-destdir=DIR\fP" 
Specify a destination directory for the download.  Defaults to the current directory.
.IP "\fB\-\-urls\fP"
Instead of downloading RPMs, list the URLs that would be downloaded.
.IP "\fB\-\-resolve\fP"
When downloading RPMs, resolve dependencies and also download the required packages.
.IP "\fB\-\-source\fP"
Instead of downloading the binary RPMs, download the source RPMs.
.IP "\fB\-e REPO, \-\-enablerepo=REPO\fP"
Enable a repository.  Using this option disables all repositories not explicitly
enabled with \-\-enablerepo.  This option may be used multiple times.  

.SH "EXAMPLES"
.IP "Download the kernel RPM to /var/tmp:"
\fByumdownloader --destdir /var/tmp kernel\fP
.IP "List the URL for the kernel and kernel-smp RPMs:"
\fByumdownloader --urls kernel kernel-smp\fP 
.PP 
.SH "FILES"
As yumdownloader uses YUM libraries for retrieving all the information, it
relies on YUM configuration for its default values like which repositories
to use. Consult YUM documentation for details:
.PP
.nf 
/etc/yum.conf
/etc/yum/repos.d/
/var/cache/yum/
.fi 

.PP 
.SH "SEE ALSO"
.nf
.I yum.conf (5)
.\"http://linux.duke.edu/yum-utils/
http://linux.duke.edu/yum/
.fi 

.PP 
.SH "AUTHORS"
.nf 
See the Authors file included with this program.
.fi 

.PP 
.SH "BUGS"
There of course aren't any bugs, but if you find any, they should be sent
to the mailing list: yum@lists.linux.duke.edu or filed in bugzilla.
.fi


ChangeLog
2007-03-20 08:03  timlau

	* yum-utils.spec: bumped yum-utils version to 1.0.4

2007-03-20 07:57  timlau

	* package-cleanup.py, repo-graph.py, repo-rss.py, repoclosure.py,
	  repoquery.py, reposync.py, repotrack.py, yumdownloader.py: Made
	  the utils work with plugins, else they wont work with RHEL5,
	  where the repos is set by a plugin.

2007-03-13 11:50  timlau

	* plugins/skip-broken/: skip-broken.conf, skip-broken.py: Added
	  'check_always' option to the skip-broken plugin.  Setting
	  'check_always=1' will make the skip-broken check run every time.
	  (rhbz #231523) Also merged changes from yum-utils HEAD

2007-02-19 15:27  pmatilai

	* plugins/kernel-module/kernel-module.py: - fix deprecation warning
	  (rhbz#222947)

2007-02-19 10:00  timlau

	* Makefile, yum-utils.spec: - Made 'make release' work with
	  %{?dist} tag in spec.  - 'make ChangeLog' will follow CVS HEAD.
	  - Added 'archive' dependency to 'make srpm' - Added %{?dist} to
	  spec file again.

2007-02-19 08:11  timlau

	* ChangeLog, Makefile: updated 'make ChangeLog' to follow the
	  'yum-utils-1_0_X' branch only.  Rebuild ChangeLog.

2007-02-19 07:57  timlau

	* yum-utils.spec: Updated changelog in spec file

2007-02-19 07:52  timlau

	* yum-utils.spec: removed %{?dist} tag from release in spec ,
	  because it breaks the 'make release'

2007-02-19 07:40  timlau

	* ChangeLog: updated ChangeLog

2007-02-19 07:38  timlau

	* yum-utils.spec: bumped yum-utils version to 1.0.3

2007-02-19 07:35  timlau

	* plugins/priorities/priorities.py: Make priorities plugin work
	  with multiarch repositories with different repositories.
	  (foo.i386  <> foo.x86_64) rhbz #227540

2007-02-15 07:38  timlau

	* yum-utils.spec, plugins/versionlock/versionlock.list: removed
	  installation of default versionlock.list

2007-02-14 14:31  timlau

	* repo-rss.py: Add a better formating for description, get rid of
	  the body tag for the xml generated by repo-rss.  Thanks to
	  Aurelien Bompard (rhbz #193111).

2007-02-14 11:37  timlau

	* yum-utils.spec, plugins/versionlock/versionlock.list: added
	  versionlock.list to versionlock plugin to avoid at lot of
	  bugzilla reports where people have installed all plugins and then
	  yum,pirut etc. are failing because the versionlock.list file dont
	  exist.

2007-02-14 10:54  timlau

	* yum-utils.spec: fixed skip-broken description (--ignore-broken ->
	  --skip-broken)

2007-02-14 10:45  timlau

	* repo-rss.py: Added patch to make repo-rss create one feed per
	  group found in comps.xml.  Thanks to Aurelien Bompard (rhbz
	  #191061)

2007-02-14 10:00  timlau

	* plugins/downloadonly/downloadonly.py: Added '--downloaddir'
	  option to downloadonly plugin. the option make it posible to
	  change the directory where packages are downloaded.  Thanks to
	  Dmitry Butskoy (rhbz #213623)

2007-02-08 13:39  timlau

	* yum-utils.spec: modified spec til to work with yum-changelog man
	  pages Added missing wildcards, so it can build

2007-02-08 13:25  timlau

	* yum-utils.spec: modified spec til to work with yum-changelog man
	  pages

2007-02-08 13:20  timlau

	* yum-utils.spec: modified spec til to work with yum-changelog man
	  pages

2007-02-08 13:16  timlau

	* docs/: Makefile, yum-changelog.1, yum-changelog.conf.5: Added
	  yum-changelog man pages by Chitlesh GOORAH and changed the docs
	  Makefile to install them.

2007-02-08 13:13  timlau

	* docs/: yum-changelog.1, yum-changelog.conf.5: Added yum-changelog
	  man pages by Chitlesh GOORAH and changed the docs Makefile to
	  install them.

2007-02-08 13:13  timlau

	* docs/yum-changelog.1: file yum-changelog.1 was added on branch
	  yum-utils-1_0_X on 2007-02-08 13:16:51 +0000

2007-02-08 13:13  timlau

	* docs/yum-changelog.conf.5: file yum-changelog.conf.5 was added on
	  branch yum-utils-1_0_X on 2007-02-08 13:16:51 +0000

2007-02-06 11:29  timlau

	* yum-utils.spec: Added %{?dist} to release in spec file, so the
	  spec can be used as is in Fedora Extras and it dont hurt other
	  distros.

2007-02-05 13:46  timlau

	* repoclosure.py: Added patch by Michael Schwendt  to sort
	  repoclosure output (yum bz# 681)

2007-01-25 18:31  timlau

	* yumdownloader.py: Fix traceback in yumdownloader using --resolve
	  because rpmdb was not initialized.

2007-01-24 19:16  slack

	* plugins/fedorakmod/: README, fedorakmod.py: Clean up some
	  comments and debug statements in fedorakmod.py.  Also, updated
	  the TODO list in the README.

2007-01-24 07:16  pmatilai

	* plugins/tsflags/tsflags.py: - fix deprecation warning on plugin
	  type

2007-01-23 23:17  slack

	* plugins/fedorakmod/: fedorakmod.conf, fedorakmod.py:
	  fedorakmod.conf: Turn off installallkmods by default.

	  fedorakmod.py: Work around Yum not populating the provides prco
	  information if that is the case.  With installallkmods on you can
	  type "yum install kmod-sysprof" and have that module installed
	  for all installed kernels and kernel types.

2007-01-11 19:20  pmatilai

	* yum-builddep.py: - resurrect builddep by package name (rh
	  #220842) - enable core type plugins (eg to allow for
	  fastestmirror)

2007-01-11 17:13  pmatilai

	* repoquery.py: - fix a stupid typo breaking prco operations

2007-01-03 14:18  timlau

	* Makefile: minor change in the Makefile, added Makefile depencies
	  to the upload  section.

2006-12-31 20:40  timlau

	* ChangeLog: updated ChangeLog

2006-12-31 20:40  timlau

	* yum-utils.spec: bumped yum-utils version to 1.0.2

2006-12-31 09:15  timlau

	* plugins/skip-broken/skip-broken.py: Changed the CLI option to
	  active the skip-broken from '--ignore-broken' to '--skip-broken',
	  because it make more sense.

2006-12-20 08:16  pmatilai

	* repoquery.py: - fix deprecation warning

2006-12-02 04:16  skvidal

	* docs/yumdownloader.1:
	  close rh bug # 218100 - typo in yumdownloader man page.

2006-11-13 14:50  timlau

	* plugins/versionlock/versionlock.py: fixed AttributeError:
	  YumAvailablePackageSqlite instance has no attribute 'returnNevra

yumdownloader.py
#!/usr/bin/python

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import sys

sys.path.insert(0, '/usr/share/yum-cli')

import yum
import yum.Errors
import os
import shutil
import output
import rpmUtils.arch
from urlgrabber.progress import TextMeter
import logging
from yum.packages import parsePackages
from yum.misc import getCacheDir
from optparse import OptionParser
from urlparse import urljoin

def initYum(yumconfigfile):
    global logger
    my = yum.YumBase()
    my.doConfigSetup(fn=yumconfigfile,init_plugins=True) # init yum, without plugins
    my.conf.uid = os.geteuid()
    if my.conf.uid != 0:
        cachedir = getCacheDir()
        if cachedir is None:
            logger.error("Error: Could not make cachedir, exiting")
            sys.exit(50)
        my.repos.setCacheDir(cachedir)
    my.repos.setProgressBar(TextMeter(fo=sys.stdout))
    my.repos.callback = output.CacheProgressCallback()

    return my

def parseArgs():
    usage = "usage: %s [options] package1 [package2] [package..]" % sys.argv[0]
    parser = OptionParser(usage=usage)
    parser.add_option("-c", "--config", default='/etc/yum.conf',
      help='config file to use (defaults to /etc/yum.conf)')
    parser.add_option("--destdir", default=".", dest="destdir",
      help='destination directory (defaults to current directory)')
    parser.add_option("--urls", default=False, dest="urls", action="store_true",
      help='just list the urls it would download instead of downloading')
    parser.add_option("--resolve", default=False, dest="resolve", action="store_true",
      help='resolve dependencies and download required packages')
    parser.add_option("--source", default=False, dest="source", action="store_true",
      help='operate on source packages')
    parser.add_option("-e","--enablerepo", default=[], action="append", dest="repo",
      help='enable repository')

    (opts, args) = parser.parse_args()
    if len(args) < 1: 
        parser.print_help()
        sys.exit(0)
    return (opts, args)

def main():
    global logger
    logger = logging.getLogger("yum.verbose.yumdownloader")
    (opts, args) = parseArgs()
    my = initYum(opts.config)

    if len(opts.repo) > 0:
        myrepos = []
        
        # find the ones we want
        for glob in opts.repo:
            myrepos.extend(my.repos.findRepos(glob))
        
        # disable them all
        for repo in my.repos.repos.values():
            repo.disable()
        
        # enable the ones we like
        for repo in myrepos:
            repo.enable()

    my.doRpmDBSetup()
    my.doRepoSetup()
    archlist = None
    if opts.source:
        archlist = rpmUtils.arch.getArchList() + ['src']

    my.doSackSetup(archlist=archlist)

    avail = my.pkgSack.returnPackages()

    toDownload = []

    packages = args
    for pkg in packages:
        toActOn = []
        exactmatch, matched, unmatched = parsePackages(avail, [pkg])
        installable = yum.misc.unique(exactmatch + matched)
        if len(unmatched) > 0: # if we get back anything in unmatched, it fails
            logger.error('No Match for argument %s' % pkg)
            continue

        for newpkg in installable:
            # This is to fix Bug 469
            # If there are matches to the package argument given but there
            # are no source packages, this can be caused because the
            # source rpm this is built from has a different name
            # for example: nscd is built from the glibc source rpm
            # We find this name by parsing the sourcerpm filename
            # (this is ugly but it appears to work)
            # And finding a package with arch src and the same
            # ver and rel as the binary package
            # That should be the source package
            # Note we do not use the epoch to search as the epoch for the
            # source rpm might be different from the binary rpm (see
            # for example mod_ssl)
            if opts.source and newpkg.arch != 'src':
                name = newpkg.returnSimple('sourcerpm').rsplit('-',2)[0]
                src = my.pkgSack.searchNevra(name=name, arch = 'src',
                  ver = newpkg.version,
                  rel = newpkg.release
                )
                toActOn.extend(src)
            else:
                toActOn.append(newpkg)

        if toActOn:
            if opts.source:
                toDownload.extend(my.bestPackagesFromList(toActOn, 'src'))
            else:
                toDownload.extend(my.bestPackagesFromList(toActOn))
    
    # If the user supplies to --resolve flag, resolve dependencies for
    # all packages
    # note this might require root access because the headers need to be
    # downloaded into the cachedir (is there a way around this)
    if opts.resolve:
        my.doTsSetup()
        my.localPackages = []
        # Act as if we were to install the packages in toDownload
        for po in toDownload:
            my.tsInfo.addInstall(po)
            my.localPackages.append(po)
        # Resolve dependencies
        my.resolveDeps()
        # Add newly added packages to the toDownload list
        for pkg in my.tsInfo.getMembers():
            if not pkg in toDownload:
                toDownload.append(pkg)
        
    for pkg in toDownload:
        n,a,e,v,r = pkg.pkgtup
        packages =  my.pkgSack.searchNevra(n,e,v,r,a)
        for download in packages:
            repo = my.repos.getRepo(download.repoid)
            remote = download.returnSimple('relativepath')
            if opts.urls:
                url = urljoin(repo.urls[0],remote)
                logger.info('%s' % url)
                continue
            local = os.path.basename(remote)
            local = os.path.join(opts.destdir, local)
            if (os.path.exists(local) and 
                str(os.path.getsize(local)) == download.returnSimple('packagesize')):
                logger.error("%s already exists and appears to be complete" % local)
                continue
            # Disable cache otherwise things won't download
            repo.cache = 0
            download.localpath = local # Hack: to set the localpath we want.
            path = repo.getPackage(download)

            if not os.path.exists(local) or not os.path.samefile(path, local):
                progress = TextMeter()
                progress.start(basename=os.path.basename(local),
                               size=os.stat(path).st_size)
                shutil.copy2(path, local)
                progress.end(progress.size)

if __name__ == '__main__':
    main()
